### ISSUE TITLE
length assertions incorrectly mark code as unreachable

### ISSUE THREAD

**Issue by marcofrasvda** [astral-sh/ty#3710]
### Summary

I have a long test that checks that a dictionary is correctly updated after each interaction, and after the second assertion everything becomes unreachable. Here's a repro:
```py
class EventHandler:
    def __init__(self):
        self.events: dict[str, str] = {}

    def add(self, id: str, name: str):
        self.events[id] = name

evh = EventHandler()
assert len(evh.events) == 0
evh.add("ab123", "test")
assert len(evh.events) == 1
# After the assertion above, everything is unreachable
assert evh.events["ab123"] == "test"
```

The exact message on hover:
```
Code is unreachable
This may depend on your current environment and settings (ty)
```

This is pretty annoying since it seems some syntax highlighting and most autocomplete are disabled in unreachable code, which is perhaps intended but it means I have to comment out assertions to work on this code. I guess one issue might be that ty sees that the length of the dictionary is zero, and considers the second assertion always false since 0 != 1.

### Version

0.0.46

**Comment by carljm** [comment]
> I guess one issue might be that ty sees that the length of the dictionary is zero, and considers the second assertion always false since 0 != 1.

Yes, that is exactly the issue. We may need to apply this narrowing only to immutable types (basically tuples). Detecting mutations accurately seems quite difficult.

**Comment by carljm** [comment]
@charliermarsh @AlexWaygood you worked on the length narrowing PR -- did this come up at all in discussion there? Any ideas of what we could do that's less of a blunt instrument than "refuse to apply this narrowing to any type we don't know is immutable"?

**Comment by charliermarsh** [comment]
Early implementations of this PR attempted to do that https://github.com/astral-sh/ruff/pull/25347 (and you can see it in the PR summary history); I'll also link you to some internal threads on Slack.
